[SourceKit] Make sure we reuse ASTContext in function bodies for solver-based cursor info

ahoppen opened the pull request:
The main problem that prevented us from reusing the ASTContext was that we weren’t remapping the `LocToResolve` in the temporary buffer that only contains the re-parsed function back to the original buffer. Thus `NodeFinder` couldn’t find the node that we want to get cursor info for.

Getting AST reuse to work for top-level items is harder because it currently heavily relies on the `HasCodeCompletion` state being set on the parser result. I’ll try that in a follow-up PR.

rdar://103251263

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[SourceKit] Sort ReferencedDecls when printing them

This assures stable sort order when comparing solver-based and AST-based results.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[SourceKit] Disable completion-like cursor info for ParamDecls

ahoppen committed:
[Sema] Pass self nominal type as a parameter to `CheckRedeclarationRequest`

The self nominal type decl needs to be passed as a parameter in here so this request doesn't tigger self nominal type computation, which results in an assertion failure.

ahoppen committed:
[IDE] Type check property wrappers so their USRs can be reported in cursor info

ahoppen committed:
[SymbolGraph] Sort edges to get consistent symbol graph string representations

ahoppen commented:
The SourceKit failure `ResultTests.swift:concurrent-834:710` is an improvement. In the following reduced reproducer the solver-based implementation is correctly reporting the shorthand shadow. The AST-based implementation is not

```swift
extension Invalid {
  func tryIsSuccess(_ text: String?) {
    // RUN: %sourcekitd-test -req=cursor -pos=%(line + 1):15 %s -- %s
    guard let text else {}
  }
}
```

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[SourceKit] Trigger generation of synthesized accessors in solver-based cursor info

ahoppen commented:
@swift-ci Please SourceKit stress test

rintaro commented on lib/IDETool/IDEInspectionInstance.cpp:
```diff
@@ -331,6 +334,17 @@ bool IDEInspectionInstance::performCachedOperationIfPossible(
                            1);
     SM.setIDEInspectionTarget(newBufferID, newOffset);
 
+    MapSourceLocIntoOriginalBuffer = [&SM, startOffset, tmpBufferID,
+                                      newBufferID](SourceLoc Loc) -> SourceLoc {
+      if (SM.findBufferContainingLoc(Loc) == newBufferID) {
+        return SM.getLocForOffset(tmpBufferID,
+                                  SM.getLocOffsetInBuffer(Loc, newBufferID) +
+                                      startOffset);
+      } else {
+        return Loc;
+      }
+    };
+
```
AFAICT, `tmpBufferID` is a buffer ID in `tmpSM`, not `SM`. So I believe `SM.getLocForOffset(tmpBufferID, ...)` is not correct, and this PR currently happens to work because the bufferID is the same `1`.

Also, I think this whole `SourceLoc` conversion mechanism should be handled in `SourceManager`.
I think we need to:
a) get offset, line, and column in the original buffer
b) compare source loc between beyond buffer IDs.

For (a) line and columns can be retrieved by `getPresumedLineAndColumnForLoc`. For `offset`, I would add an optional parameter to `SourceManager::openVirtualFile()` like `offsetOffset` (I'm serious) and remember it in `VirtualFiles`, then create `SourceManager::getPresumedLocOffsetInBuffer()` so retrieve the adjusted "offset".

For (b), `SourceManager` knows the replaced source ranges, in `ReplacedRanges`, so it should be able to "compare" the locations beyond source buffer if the loc is a part of `ReplacedRanges` (like `ASTScope` does).

Could you explore this option? I feel passing around this `MapSourceLocIntoOriginalBuffer` is a too much burden.

rintaro reviewed (changes requested):
The divergence between "cursor info" and other requests in `IDEInspectionInstance`  is getting bigger.

* Existence of code completion _token_
* Callbacks requirement
* `IgnoreSwiftSourceInfo` requirement
* `SourceLoc` requirements

I feel passing different constant arguments depending on the request kind is not great for code readability.

We might want to reconsider whether we should continue to use a `IDEInspectionInstance` for everything. Maybe it would be another `IDEInspectionInstance` instance with different option, or different "cursor info instance" sharing some code (e.g. dependency checking) with `IDEInspectionInstance`.

ahoppen committed:
[SourceKit] Make sure we reuse ASTContext in function bodies for solver-based cursor info

The main problem that prevented us from reusing the ASTContext was that we weren’t remapping the `LocToResolve` in the temporary buffer that only contains the re-parsed function back to the original buffer. Thus `NodeFinder` couldn’t find the node that we want to get cursor info for.

Getting AST reuse to work for top-level items is harder because it currently heavily relies on the `HasCodeCompletion` state being set on the parser result. I’ll try that in a follow-up PR.

rdar://103251263

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[sourcekitd-test] Flush stdout and stderr before executing a shell command

Otherwise the order of sourcekitd-test’s output and the shell command is undefined.

ahoppen committed:
[sourcekitd-test] Don't cache memory buffers

If we cache the buffers between multiple requests we don’t pick up if the source file has changed on disk since we last read it.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test macOS

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[CursorInfo] Fix a bug that caused incorrect discriminators to be assigned to values in completion-like cursor info

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
Note to self: All cursor info verification failures in the stress tester are improvement. Most of them are because we now correctly return an optional and a non-optional result for an incomplete `if let`. Others are cases where the USR previously contained error types but no longer does.

ahoppen commented:
@swift-ci Please smoke test
